File: priv/native/android/src/main/java/mob/ble/MobFetchProtocol.kt
package mob.ble

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.charset.StandardCharsets

object MobFetchProtocol {
    private val REQUEST_MAGIC = byteArrayOf('M'.code.toByte(), 'F'.code.toByte(), 'Q'.code.toByte())
    private val RESPONSE_MAGIC = byteArrayOf('M'.code.toByte(), 'F'.code.toByte(), 'R'.code.toByte())
    const val VERSION = 1
    const val STATUS_OK = 0
    const val STATUS_NOT_FOUND = 1
    const val STATUS_INVALID_REQUEST = 2

    fun statusName(status: Int): String =
        when (status) {
            STATUS_OK -> "ok"
            STATUS_NOT_FOUND -> "not_found"
            STATUS_INVALID_REQUEST -> "invalid_request"
            else -> "unknown"
        }

    data class Request(
        val requestId: String,
        val messageIdHash: ByteArray,
        val requesterPeerId: String?
    )

    data class Response(
        val requestId: String,
        val messageIdHash: ByteArray,
        val status: Int,
        val envelope: ByteArray?,
        val reason: String?
    )

    fun encodeRequest(request: Request): ByteArray {
        require(request.requestId.isNotEmpty())
        require(request.messageIdHash.size == 8)
        val requestId = request.requestId.toByteArray(StandardCharsets.UTF_8)
        val requester = request.requesterPeerId?.toByteArray(StandardCharsets.UTF_8) ?: ByteArray(0)
        require(requestId.size <= 255)
        require(requester.size <= 255)

        return ByteArrayOutputStream().use { out ->
            out.write(REQUEST_MAGIC)
            out.write(VERSION)
            out.write(requestId.size)
            out.write(requestId)
            out.write(request.messageIdHash)
            out.write(requester.size)
            out.write(requester)
            out.toByteArray()
        }
    }

    fun decodeRequest(bytes: ByteArray): Request? {
        if (bytes.size < 3 + 1 + 1 + 8 + 1) return null
        if (!bytes.copyOfRange(0, 3).contentEquals(REQUEST_MAGIC)) return null
        if ((bytes[3].toInt() and 0xFF) != VERSION) return null
        var offset = 4
        val requestIdLength = bytes[offset].toInt() and 0xFF
        offset += 1
        if (offset + requestIdLength + 8 + 1 > bytes.size) return null
        val requestId = String(bytes, offset, requestIdLength, StandardCharsets.UTF_8)
        offset += requestIdLength
        val hash = bytes.copyOfRange(offset, offset + 8)
        offset += 8
        val requesterLength = bytes[offset].toInt() and 0xFF
        offset += 1
        if (offset + requesterLength > bytes.size) return null
        val requester = if (requesterLength == 0) null else String(bytes, offset, requesterLength, StandardCharsets.UTF_8)
        if (requestId.isEmpty()) return null
        return Request(requestId, hash, requester)
    }

    fun encodeResponse(response: Response): ByteArray {
        require(response.requestId.isNotEmpty())
        require(response.messageIdHash.size == 8)
        val requestId = response.requestId.toByteArray(StandardCharsets.UTF_8)
        val payload = response.envelope ?: response.reason?.toByteArray(StandardCharsets.UTF_8) ?: ByteArray(0)
        require(requestId.size <= 255)
        require(payload.size <= 65535)

        return ByteArrayOutputStream().use { out ->
            out.write(RESPONSE_MAGIC)
            out.write(VERSION)
            out.write(response.status)
            out.write(requestId.size)
            out.write(requestId)
            out.write(response.messageIdHash)
            out.write(ByteBuffer.allocate(2).order(ByteOrder.BIG_ENDIAN).putShort(payload.size.toShort()).array())
            out.write(payload)
            out.toByteArray()
        }
    }

    fun decodeResponse(bytes: ByteArray): Response? {
        if (bytes.size < 3 + 1 + 1 + 1 + 8 + 2) return null
        if (!bytes.copyOfRange(0, 3).contentEquals(RESPONSE_MAGIC)) return null
        if ((bytes[3].toInt() and 0xFF) != VERSION) return null
        val status = bytes[4].toInt() and 0xFF
        if (status !in STATUS_OK..STATUS_INVALID_REQUEST) return null
        var offset = 5
        val requestIdLength = bytes[offset].toInt() and 0xFF
        offset += 1
        if (offset + requestIdLength + 8 + 2 > bytes.size) return null
        val requestId = String(bytes, offset, requestIdLength, StandardCharsets.UTF_8)
        offset += requestIdLength
        val hash = bytes.copyOfRange(offset, offset + 8)
        offset += 8
        val payloadLength = ByteBuffer.wrap(bytes, offset, 2).order(ByteOrder.BIG_ENDIAN).short.toInt() and 0xFFFF
        offset += 2
        if (offset + payloadLength > bytes.size) return null
        val payload = bytes.copyOfRange(offset, offset + payloadLength)
        if (requestId.isEmpty()) return null
        return if (status == STATUS_OK) {
            Response(requestId, hash, status, payload, null)
        } else {
            Response(requestId, hash, status, null, String(payload, StandardCharsets.UTF_8))
        }
    }
}

File: priv/native/ios/MobFetchProtocol.swift
import Foundation

/// Swift mirror of Android's `FetchProtocol`. Same wire format —
/// MFQ/MFR magic, version 1, length-prefixed strings, big-endian payload
/// size — so the iOS requester can talk to the Android responder
/// implemented in the Android fetch Gatt.
public enum FetchProtocol {
    public static let version: UInt8 = 1
    public static let statusOK: UInt8 = 0
    public static let statusNotFound: UInt8 = 1
    public static let statusInvalidRequest: UInt8 = 2

    private static let requestMagic = Data([0x4D, 0x46, 0x51]) // "MFQ"
    private static let responseMagic = Data([0x4D, 0x46, 0x52]) // "MFR"

    public struct Request: Equatable {
        public var requestId: String
        public var messageIdHash: Data
        public var requesterPeerId: String?

        public init(requestId: String, messageIdHash: Data, requesterPeerId: String?) {
            self.requestId = requestId
            self.messageIdHash = messageIdHash
            self.requesterPeerId = requesterPeerId
        }
    }

    public struct Response: Equatable {
        public var requestId: String
        public var messageIdHash: Data
        public var status: UInt8
        public var envelope: Data?
        public var reason: String?

        public init(
            requestId: String,
            messageIdHash: Data,
            status: UInt8,
            envelope: Data?,
            reason: String?
        ) {
            self.requestId = requestId
            self.messageIdHash = messageIdHash
            self.status = status
            self.envelope = envelope
            self.reason = reason
        }
    }

    public static func encodeRequest(_ request: Request) -> Data? {
        guard !request.requestId.isEmpty,
              request.messageIdHash.count == 8 else {
            return nil
        }
        let requestId = Data(request.requestId.utf8)
        let requester = request.requesterPeerId.map { Data($0.utf8) } ?? Data()
        guard requestId.count <= 255, requester.count <= 255 else { return nil }

        var out = Data()
        out.append(requestMagic)
        out.append(version)
        out.append(UInt8(requestId.count))
        out.append(requestId)
        out.append(request.messageIdHash)
        out.append(UInt8(requester.count))
        out.append(requester)
        return out
    }

    /// Decode a peer's MFQ Request. Used by the responder
    /// (FetchGattResponder) to parse what a client wrote.
    public static func decodeRequest(_ bytes: Data) -> Request? {
        guard bytes.count >= 3 + 1 + 1 + 8 + 1 else { return nil }
        let start = bytes.startIndex
        guard bytes[start..<start.advanced(by: 3)] == requestMagic else { return nil }
        guard bytes[start.advanced(by: 3)] == version else { return nil }

        var offset = start.advanced(by: 4)
        let requestIdLength = Int(bytes[offset])
        offset = offset.advanced(by: 1)
        guard offset.advanced(by: requestIdLength + 8 + 1) <= bytes.endIndex else { return nil }

        let requestIdRange = offset..<offset.advanced(by: requestIdLength)
        guard let requestId = String(data: bytes[requestIdRange], encoding: .utf8),
              !requestId.isEmpty else { return nil }
        offset = offset.advanced(by: requestIdLength)

        let hash = Data(bytes[offset..<offset.advanced(by: 8)])
        offset = offset.advanced(by: 8)

        let requesterLength = Int(bytes[offset])
        offset = offset.advanced(by: 1)
        guard offset.advanced(by: requesterLength) <= bytes.endIndex else { return nil }

        let requesterPeerId: String?
        if requesterLength == 0 {
            requesterPeerId = nil
        } else {
            let requesterRange = offset..<offset.advanced(by: requesterLength)
            requesterPeerId = String(data: bytes[requesterRange], encoding: .utf8)
        }

        return Request(
            requestId: requestId,
            messageIdHash: hash,
            requesterPeerId: requesterPeerId
        )
    }

    /// Encode an MFR Response. Used by the responder to produce the
    /// bytes a client will read.
    public static func encodeResponse(_ response: Response) -> Data {
        precondition(!response.requestId.isEmpty)
        precondition(response.messageIdHash.count == 8)
        let requestId = Data(response.requestId.utf8)
        let payload = response.envelope
            ?? response.reason.map { Data($0.utf8) }
            ?? Data()
        precondition(requestId.count <= 255)
        precondition(payload.count <= 0xFFFF)

        var out = Data()
        out.append(responseMagic)
        out.append(version)
        out.append(response.status)
        out.append(UInt8(requestId.count))
        out.append(requestId)
        out.append(response.messageIdHash)
        // Big-endian 16-bit payload length — matches Android's
        // ByteBuffer.allocate(2).order(BIG_ENDIAN).putShort(...).
        out.append(UInt8((payload.count >> 8) & 0xFF))
        out.append(UInt8(payload.count & 0xFF))
        out.append(payload)
        return out
    }

    public static func decodeResponse(_ bytes: Data) -> Response? {
        guard bytes.count >= 3 + 1 + 1 + 1 + 8 + 2 else { return nil }
        let start = bytes.startIndex
        guard bytes[start..<start.advanced(by: 3)] == responseMagic else { return nil }
        guard bytes[start.advanced(by: 3)] == version else { return nil }
        let status = bytes[start.advanced(by: 4)]
        guard status <= statusInvalidRequest else { return nil }

        var offset = start.advanced(by: 5)
        let requestIdLength = Int(bytes[offset])
        offset = offset.advanced(by: 1)
        guard offset.advanced(by: requestIdLength + 8 + 2) <= bytes.endIndex else { return nil }

        let requestIdRange = offset..<offset.advanced(by: requestIdLength)
        guard let requestId = String(data: bytes[requestIdRange], encoding: .utf8),
              !requestId.isEmpty else { return nil }
        offset = offset.advanced(by: requestIdLength)

        let hash = Data(bytes[offset..<offset.advanced(by: 8)])
        offset = offset.advanced(by: 8)

        let payloadLength =
            (Int(bytes[offset]) << 8) | Int(bytes[offset.advanced(by: 1)])
        offset = offset.advanced(by: 2)
        guard offset.advanced(by: payloadLength) <= bytes.endIndex else { return nil }
        let payload = Data(bytes[offset..<offset.advanced(by: payloadLength)])

        if status == statusOK {
            return Response(
                requestId: requestId,
                messageIdHash: hash,
                status: status,
                envelope: payload,
                reason: nil
            )
        } else {
            let reason = String(data: payload, encoding: .utf8) ?? ""
            return Response(
                requestId: requestId,
                messageIdHash: hash,
                status: status,
                envelope: nil,
                reason: reason
            )
        }
    }
}
